src/cpp/server/server_cc.cc

```diff
@@ -1333,28 +1270,13 @@ void Server::ShutdownInternal(gpr_timespec deadline) {
     value->Wait();
   }
 
-  // Wait for all outstanding callback requests to complete
-  // (whether waiting for a match or already active).
-  // We know that no new requests will be created after this point
-  // because they are only created at server startup time or when
-  // we have a successful match on a request. During the shutdown phase,
-  // requests that have not yet matched will be failed rather than
-  // allowed to succeed, which will cause the server to delete the
-  // request and decrement the count. Possibly a request will match before
-  // the shutdown but then find that shutdown has already started by the
-  // time it tries to register a new request. In that case, the registration
-  // will report a failure, indicating a shutdown and again we won't end
-  // up incrementing the counter.
-  {
-    grpc::internal::MutexLock cblock(&callback_reqs_mu_);
-    callback_reqs_done_cv_.WaitUntil(
-        &callback_reqs_mu_, [this] { return callback_reqs_outstanding_ == 0; });
-  }
+  // Drop the shutdown ref and wait for all other refs to drop as well.
+  UnrefAndWaitLocked();
 
   // Shutdown the callback CQ. The CQ is owned by its own shutdown tag, so it
   // will delete itself at true shutdown.
-  if (callback_cq_ != nullptr) {
-    callback_cq_->Shutdown();
+  if (callback_cq_.load() != nullptr) {
```

Comment:
For my education: is it possible to have multiple threads run `Server::ShutdownInternal()` or is that illegal?  If the latter, I would suggest using acquire here.  I would presume it's not the former, or otherwise we need to use exchange here.

Reply:
The beginning of the function checks (under lock) if the shutdown_ variable is set, which causes an early exit, and then sets the shutdown_ variable otherwise. So we'll only have 1 thread making it to this point. I guess this means we can use relaxed here?

Reply:
Oh sorry, I missed that one.